### PR TITLE
Counterexample generation and checking

### PR DESCRIPTION
**Pull request by msooseth**
## Description

This adds a preliminary counterexample, `Cex` (prounounced KEX) generation. It uses the `smt2-parser` Haskell library, which necessitated a new pin on the `nixpkgs.url` and consequently, `flake.lock`.

Feedback welcome :)

## Checklist

- [x] tested locally
- [x] added automated tests
- [ ] updated the docs
- [ ] updated the changelog

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
   inputs = {
     flake-utils.url = "github:numtide/flake-utils";
-    nixpkgs.url = "github:nixos/nixpkgs/nixos-22.05";
```
Comment (arcz) [inline flake.nix:6]: Should we use the `haskell-updates` branch instead? I think it's clearer and the exact commit is pinned already in `flake.lock`.

Comment (msooseth) [reply flake.nix:6]: I'm happy with that, actually. It's the same commit, that's correct. Let me fix it to `haskell-updates`

Comment (msooseth) [reply flake.nix:6]: Cool, fixed now, I'm resolving. Please re-open if you feel like the fix was not what you expected... I hope it's good :) Thanks again!

File: src/hevm/src/EVM/SMT.hs [hunk 2]
```diff
@@ -35,16 +40,50 @@ import System.Process (createProcess, cleanupProcess, proc, ProcessHandle, std_i
...
+    mempty = CexVars
+      { calldataV = mempty
+      , storageV = mempty
+      , blockContextV = mempty
+      , txContextV = mempty
+      }
+
+data SMTCex = SMTCex
+  { calldata :: Map Text Language.SMT2.Parser.SpecConstant
+  , storage :: Text
```
Comment (zoep) [inline src/hevm/src/EVM/SMT.hs:70]: `storage`, `blockContext` and `txContext` are not currently used and I think it's fine to leave them as placeholders and add this functionality later.  I'm puzzled by their types though as their values is of `Text` type and not of `SpecConstant` type as calldata

**Review by zoep — APPROVED** [review]
Nice PR! Made two small comments but it looks good to me overall.

File: src/hevm/src/EVM/SMT.hs (hunk 2 above)
Comment (msooseth) [reply src/hevm/src/EVM/SMT.hs:70]: PR is here: https://github.com/ethereum/hevm/pull/62

File: src/hevm/src/EVM/SMT.hs [hunk 3]
```diff
@@ -692,6 +733,17 @@ withSolvers solver count cont = do
       -- put the instance back in the list of available instances
       writeChan availableInstances inst
 
+hexChar :: Char -> Int
+hexChar ch = Data.Maybe.fromMaybe (error $ "illegal char " ++ [ch]) $
+    List.elemIndex ch "0123456789abcdef"
+
+parseHex :: String -> Integer
+parseHex hex = List.foldl' f (0 ::Integer) hex where
+    f n c = (16 :: Integer)*n + toInteger (hexChar c)
```
[the remark this answers is not in the document]
Comment (msooseth) [reply src/hevm/src/EVM/SMT.hs:742]: Good catch" Thank you, I didn't know :( PR to fix is here: https://github.com/ethereum/hevm/pull/62

**Review by msooseth — COMMENTED** [review]
Sorry, I saw these too late :( I created a PR to fix them up:

https://github.com/ethereum/hevm/pull/62

**Pull request merged**